### PR TITLE
sys/ztimer: add ZTIMER_SEC

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

Wired up `ZTIMER_SEC` to the existing RTC backend, or RTT backend, or `periph_timer` backend (in this order of preference).

### Testing procedure

- `tests/ztimer_msg` has been updated to allow testing `ZTIMER_SEC`. Note: Uncomment a line in the Makefile as documented there
- `tests/ztimer_underflow` has been updated to allow testing `ZTIMER_SEC`
     - Run `make TEST_ZTIMER_CLOCK=ZTIMER_SEC -C tests/ztimer_underflow flash test`
     - This fails for me, which proves two things: a) wiring up the RTC worked, and b) there is a preexisting bug in `master` unrelated to the PR that needs fixing.

### Issues/PRs references

None

### CODE REVIEW

**Comment by maribu** [comment]
@kfessel might want to have a look at this

File: sys/ztimer/auto_init.c [hunk 1]
```diff
@@ -98,6 +100,27 @@ ztimer_clock_t *const ZTIMER_MSEC_BASE = &_ztimer_periph_timer_usec.super;
...
+#    define ZTIMER_SEC_CONVERT_LOWER        (&_ztimer_periph_timer_rtt_msec)
+#    define ZTIMER_SEC_CONVERT_LOWER_FREQ   RTT_FREQUENCY
+#  elif MODULE_ZTIMER_USEC
+static ztimer_convert_frac_t _ztimer_convert_frac_sec;
+ztimer_clock_t *const ZTIMER_SEC = &_ztimer_convert_frac_sec.super.super;
+ztimer_clock_t *const ZTIMER_SEC_BASE = &_ztimer_periph_timer_usec.super;
+#    define ZTIMER_SEC_CONVERT_LOWER        ZTIMER_USEC_BASE
+#    define ZTIMER_SEC_CONVERT_LOWER_FREQ   CONFIG_ZTIMER_USEC_BASE_FREQ
+#  else
+#  error No suitable ZTIMER_MSEC config. Maybe add USEMODULE += ztimer_usec?
```
Comment (leandrolanzieri) [inline sys/ztimer/auto_init.c:120]:
```suggestion
#  error No suitable ZTIMER_SEC config. Maybe add USEMODULE += ztimer_usec?
```

File: sys/ztimer/auto_init.c [hunk 2]
```diff
@@ -98,6 +100,27 @@ ztimer_clock_t *const ZTIMER_MSEC_BASE = &_ztimer_periph_timer_usec.super;
...
 
+#if MODULE_ZTIMER_SEC
+#  if MODULE_ZTIMER_PERIPH_RTC
+static ztimer_periph_rtc_t _ztimer_periph_timer_rtc_sec;
+ztimer_clock_t *const ZTIMER_SEC = &_ztimer_periph_timer_rtc_sec;
+#  elif MODULE_ZTIMER_MSEC
+static ztimer_convert_frac_t _ztimer_convert_frac_sec;
+ztimer_clock_t *const ZTIMER_SEC = &_ztimer_convert_frac_sec.super.super;
+ztimer_clock_t *const ZTIMER_SEC_BASE = &_ztimer_periph_timer_msec.super;
+#    define ZTIMER_SEC_CONVERT_LOWER        (&_ztimer_periph_timer_rtt_msec)
```
Comment (leandrolanzieri) [inline sys/ztimer/auto_init.c:111]: is it ok to use this symbol here? doesn't it need the rtt?

**Comment by kaspar030** [comment]
Long overdue! But I think we should try to not use RTC. On most platforms, it shares the same hardware with RTT anyways (requiring to chose one). Also the RTC interface is clumsy to use if only a seconds granularity counter is desired (like in this case).

So, please change to use RTT as preference (possibly a converted ZTIMER_MSEC), and only fall back to RTC if no RTT is available.

**Comment by kaspar030** [comment]
>     * This fails for me, which proves two things: a) wiring up the RTC worked, and b) there is a preexisting bug in `master` unrelated to the PR that needs fixing.

Yeah, the RTC backend probably doesn't yet deal with setting 0 or 1 properly. It would need to set at least +2 to be safe.

**Comment by maribu** [comment]
> So, please change to use RTT as preference (possibly a converted ZTIMER_MSEC), and only fall back to RTC if no RTT is available.

Hmm, if one pulls in `ztimer_periph_rtc`, IMO we really should use this for `ZTIMER_SEC`. So far there is no logic that every pulls in `ztimer_periph_rtc`, so it is up to the user to add this module. And for the reasons you provided against using the RTC, I don't think we should automatically pull it in.

Btw.: Shouldn't we (unrelated of this PR) use `FEATURES_OPTIONAL += periph_rtt` for `ztimer_msec` and `ztimer_sec` and automatically use `ztimer_periph_rtt` if `periph_rtt` is used?

File: sys/ztimer/auto_init.c (hunk 2 above)
Comment (maribu) [reply sys/ztimer/auto_init.c:111]: Good catch. I changed the preprocessor conditional to only uses this if both `ztimer_msec` and `ztimer_periph_rtt` are used.

Maybe it makes sense to decouple this from `ztimer_msec` to also profit from the power savings when `ztimer_sec` but not `ztimer_msec` is used. I'll fix this.

**Comment by maribu** [comment]
Oh, I chose the wrong commit with `--fixup=`. I hope I remember to manually squash to fix this :-D

File: sys/ztimer/auto_init.c (hunk 2 above)
Comment (maribu) [reply sys/ztimer/auto_init.c]: Done. RTT is now used for `ZTIMER_SEC` when available regardless of `ZTIMER_MSEC`, unless the application/user explicitly pulled in `ztimer_periph_rtc`.

**Comment by kaspar030** [comment]
> Btw.: Shouldn't we (unrelated of this PR) use `FEATURES_OPTIONAL += periph_rtt` for `ztimer_msec` and `ztimer_sec` and automatically use `ztimer_periph_rtt` if `periph_rtt` is used?

That's an FAQ by now. The issue is that if ztimer_periph_rtt is used, it'll be used for ztimer_msec. But not all periph_rtt can actually do 1000 or 1024 Hz (or are configured that way). I think we need to add more fine-grained rtt capabilities as features.

**Comment by maribu** [comment]
>  I think we need to add more fine-grained rtt capabilities as features.

:+1:

**Review by kfessel — CHANGES_REQUESTED** [review]
I did not want to respond without measuring. Yesterday i found my nucleo-l073rz having a IDD bride. I tested different configuration (non extra, rtt, rtc, rtt and rtc). Not every configuration could be built first I fixed that, patch attached an added the pm_layerd configuration for ZTIMER_SEC

**Comment by maribu** [comment]
@kfessel: Care to also squash when rebasing?

File: sys/ztimer/auto_init.c [hunk 3]
```diff
@@ -166,4 +196,22 @@ void ztimer_init(void)
...
+    LOG_DEBUG("ztimer_init(): ZTIMER_SEC convert_frac from %lu to 1\n",
+              (long unsigned)ZTIMER_SEC_CONVERT_LOWER_FREQ);
+    ztimer_convert_frac_init(&_ztimer_convert_frac_sec,
+                             ZTIMER_SEC_CONVERT_LOWER,
+                             FREQ_1HZ, ZTIMER_SEC_CONVERT_LOWER_FREQ);
+#  endif
+#  ifdef MODULE_PM_LAYERED
+    LOG_DEBUG("ztimer_init(): ZTIMER_SEC setting required_pm_mode to %i\n",
+              CONFIG_ZTIMER_SEC_REQUIRED_PM_MODE);
+    ZTIMER_SEC->required_pm_mode = CONFIG_ZTIMER_SEC_REQUIRED_PM_MODE;
```
Comment (maribu) [inline sys/ztimer/auto_init.c:214]: @kfessel: Can you take a look at https://github.com/RIOT-OS/RIOT/pull/15911 I think this needs to be applied to the hardware clock. In case of the RTC, this is already the hardware clock, but in case of RTT, it should be applied to the RTT clock sitting below the convert frac clock.

**Comment by kfessel** [comment]
my Test Example 

[ztimer_pm.zip](https://github.com/RIOT-OS/RIOT/files/5933543/ztimer_pm.zip)

File: sys/ztimer/auto_init.c (hunk 3 above)
Comment (kfessel) [reply sys/ztimer/auto_init.c:214]: you are right and while writing these lines i thought this is wrong but its the current way to go and i was already searching if there is an issue or PR in this direction.
so these PRs are in "conflict". one has to merged the other has to follow

**Comment by kfessel** [comment]
with restructure for flexibility i also fixed the issue of #15911

**Comment by kfessel** [comment]
uncrustify keeps calling out:
`sys/include/ztimer.h#L295`
it thinks the * is an operation but it is a pointer-marker (i like the space after the * but it is to my understanding against the uncrustify rules)

**Comment by maribu** [comment]
> uncrustify keeps calling out:
> `sys/include/ztimer.h#L295`
> it thinks the * is an operation but it is a pointer-marker (i like the space after the * but it is to my understanding against the uncrustify rules)

Apparently the old `uncrustify` release sometimes confuses unary and binary `*` operator. The current stable release of `uncrustify` doesn't complain. So just ignore this and once the new stable release makes it to the CI, the warning will disappear.

File: sys/include/ztimer/config.h [hunk 4]
```diff
@@ -104,17 +104,37 @@ extern "C" {
...
+ * @brief   The minimum pm mode required for ZTIMER_TIMER to run.
  */
-#ifndef CONFIG_ZTIMER_USEC_REQUIRED_PM_MODE
-#define CONFIG_ZTIMER_USEC_REQUIRED_PM_MODE ZTIMER_CLOCK_NO_REQUIRED_PM_MODE
+#ifndef CONFIG_ZTIMER_TIMER_REQUIRED_PM_MODE
+    #ifdef CONFIG_ZTIMER_USEC_REQUIRED_PM_MODE
+        #define CONFIG_ZTIMER_TIMER_REQUIRED_PM_MODE CONFIG_ZTIMER_USEC_REQUIRED_PM_MODE
+    #else
+        #define CONFIG_ZTIMER_TIMER_REQUIRED_PM_MODE ZTIMER_CLOCK_NO_REQUIRED_PM_MODE
+    #endif
```
Comment (maribu) [inline sys/include/ztimer/config.h:114]:
```suggestion
#ifndef CONFIG_ZTIMER_TIMER_REQUIRED_PM_MODE
#  ifdef CONFIG_ZTIMER_USEC_REQUIRED_PM_MODE
#     define CONFIG_ZTIMER_TIMER_REQUIRED_PM_MODE CONFIG_ZTIMER_USEC_REQUIRED_PM_MODE
#  else
#    define CONFIG_ZTIMER_TIMER_REQUIRED_PM_MODE ZTIMER_CLOCK_NO_REQUIRED_PM_MODE
#  endif
```

We don't really have style of preprocessor stuff covered by the coding convention, but it is IMO better to keep the style at least consistent with the other macros in the same source file.

The reasoning for keeping the `#` as first character is that this is actually required by the C standard (at least in older versions). Spaces between the `#` and the preprocessor statement are explicitly valid and ignored. I hove no idea why two spaces are used for indent instead of the usual four, but let's just stick with that choice for consistency.

File: sys/ztimer/auto_init.c [hunk 5]
```diff
@@ -52,65 +53,150 @@
 #define FREQ_1MHZ       1000000LU
 #define FREQ_250KHZ     250000LU
 #define FREQ_1KHZ       1000LU
+#define FREQ_1HZ        1LU
+
+#if CONFIG_ZTIMER_USEC_TYPE_PERIPH_TIMER
+//#if MODULE_ZTIMER_PERIPH_TIMER
```
Comment (maribu) [inline sys/ztimer/auto_init.c:59]:
```suggestion
```

File: sys/include/ztimer/config.h (hunk 4 above)
Comment (kfessel) [reply sys/include/ztimer/config.h:114]: I will change these when the first Murdock report is ready

File: sys/ztimer/auto_init.c (hunk 5 above)
Comment (kfessel) [reply sys/ztimer/auto_init.c:59]: I will change these when the first Murdock report is ready

**Comment by kfessel** [comment]
I would like to squash my fixups.

**Comment by maribu** [comment]
> I would like to squash my fixups.

Go ahead :-)

**Comment by kfessel** [comment]
@kaspar030 @leandrolanzieri @maribu : it went green :tada:

The restructuring implements following logic of timer choice:
- if module `ztimer_usec` is used use the (xtimer) `ztimer_periph_timer` for it.
- if module `ztimer_msec` is used and `ztrimer_periph_rtt` use `ztimer_periph_rtt` for it else use the (xtimer) `ztimer_periph_timer` for it.
- if module `ztimer_sec` is used and `ztimer_periph_rtc` is used use `ztimer_periph_rtc` for it else if `ztrimer_periph_rtt` used, use `ztimer_periph_rtt` for it else use the (xtimer) `ztimer_periph_timer` for it.
- Non of the `ztimer_*sec` modules depend on each other they only depend on `ztimer_periph*` assuming `ztimer_periph_timer` is always available they can always be implemented (the: 'No suitable ZTIMER_MSEC config. Maybe add USEMODULE += ztimer_usec?` error message is gone the replacement message should never be triggert since there is always a ztimer_periph_timer atm).

File: sys/ztimer/auto_init.c [hunk 6]
```diff
@@ -52,78 +53,201 @@
...
+static ztimer_periph_rtc_t ZTIMER_RTC;
+#endif
+
+/* Step 3: setup constants for ztimers and memory for converters */
+
+#if MODULE_ZTIMER_USEC
+#  if defined(ZTIMER_TIMER)
+ztimer_clock_t *const ZTIMER_USEC_BASE = &ZTIMER_TIMER_CLK;
+#  else
+#    error No suitable ZTIMER_USEC config. No basic timer configuration missing?
```
Comment (maribu) [inline sys/ztimer/auto_init.c:150]:
```suggestion
#    error No suitable ZTIMER_USEC config. Basic timer configuration missing?
```

File: sys/ztimer/auto_init.c [hunk 7]
```diff
@@ -52,78 +53,201 @@
...
+#if MODULE_ZTIMER_SEC
+#  ifdef ZTIMER_RTC
+ztimer_clock_t *const ZTIMER_SEC_BASE = &ZTIMER_RTC_CLK;
+ztimer_clock_t *const ZTIMER_SEC = &ZTIMER_RTC_CLK;
+#  elif defined(ZTIMER_RTT)
+ztimer_clock_t *const ZTIMER_SEC_BASE = &ZTIMER_RTT_CLK;
+#  elif defined(ZTIMER_TIMER)
+ztimer_clock_t *const ZTIMER_SEC_BASE = &ZTIMER_TIMER_CLK;
+#  else
+    #  error No suitable ZTIMER_SEC config. No rtc, rtt or basic timer?
```
Comment (maribu) [inline sys/ztimer/auto_init.c:188]:
```suggestion
#  else
#    error No suitable ZTIMER_SEC config. No rtc, rtt or basic timer?
```

File: sys/ztimer/auto_init.c [hunk 8]
```diff
@@ -52,78 +53,201 @@
...
                              CONFIG_ZTIMER_USEC_BASE_FREQ,
                              WIDTH_TO_MAXVAL(CONFIG_ZTIMER_USEC_WIDTH));
+#  ifdef MODULE_PM_LAYERED
+    LOG_DEBUG("ztimer_init(): ZTIMER_TIMER setting required_pm_mode to %i\n",
+              CONFIG_ZTIMER_TIMER_REQUIRED_PM_MODE);
+    ZTIMER_TIMER_CLK.required_pm_mode = CONFIG_ZTIMER_TIMER_REQUIRED_PM_MODE;
 #  endif
+#endif
+
+#ifdef USE_ZTIMER_RTT
```
Comment (maribu) [inline sys/ztimer/auto_init.c:215]:
```suggestion
#if USE_ZTIMER_RTT
```

Maybe? This would allow defining it two zero when needed for quick and dirty debugging.

**Review by maribu — COMMENTED** [review]
Untangling ztimer backend initialization from the initialization of the default clocks did increase IMO readability quite a bit. Some minor commets, please squash right in.

**Comment by maribu** [comment]
@vincent-d, @kaspar030, @jue89: Maybe one of you could review? I have conflicts of interests here that prevent me from giving an ACK, since I started this PR.

File: sys/ztimer/auto_init.c (hunk 8 above)
Comment (kfessel) [reply sys/ztimer/auto_init.c:215]: the possibility to define it to 0 for debugging purposes is a good find and of cause the error message erorros. I will change and squash

Comment (kfessel) [reply sys/ztimer/auto_init.c:215]: and back again this will not work the way you intend to use it since `USE_ZTIMER_*` is an internal marker whether to alloc (static) space for the coresponing hardware the availability is decided by the use of the module leading to the name (ZTIMER_*) being defined.
but i change it for consitency

Comment (kfessel) [reply sys/ztimer/auto_init.c:215]: this made me think about the naming of the `USE_ZTIMER...` defines i think they should be renamed `INIT_ZTiMMER...`

Comment (kfessel) [reply sys/ztimer/auto_init.c]: DONE

**Comment by kfessel** [comment]
@vincent-d, @kaspar030, @jue89: Is there anything left to be fixed before this can be merged?

**Comment by kfessel** [comment]
> Long overdue! But I think we should try to not use RTC. On most platforms, it shares the same hardware with RTT anyways (requiring to chose one). Also the RTC interface is clumsy to use if only a seconds granularity counter is desired (like in this case).
> 
> So, please change to use RTT as preference (possibly a converted ZTIMER_MSEC), and only fall back to RTC if no RTT is available.


i changed the auto_init to prefer RTT over RTC for ZTIMER_SEC

File: sys/ztimer/auto_init.c [hunk 9]
```diff
@@ -52,78 +53,206 @@
...
 #if MODULE_ZTIMER_USEC
-#  if CONFIG_ZTIMER_USEC_TYPE_PERIPH_TIMER
-static ztimer_periph_timer_t _ztimer_periph_timer_usec = {
+#  ifndef INIT_ZTIMER_TIMER
+#    define INIT_ZTIMER_TIMER 1
+#  endif
+#endif
+
+#if MODULE_ZTIMER_MSEC
+#  if defined(ZTIMER_RTT) && RTT_FREQUENCY >= FREQ_1KHZ
```
Comment (fjmolinas) [inline sys/ztimer/auto_init.c:93]: Isn't there frequency shifting as well?

**Review by fjmolinas — COMMENTED** [review]
Generally looks good some comments though:

- it would be nice if the choice of initialization and backend was documented in another way as only with ifdefs, since they are really hard to read.
- can we add distinctive tests for the different timers Since the application is the same mostly, I think an approach as the one used for the network driver compile test can be used
- the history needs fixing :)
- there are some uncrustify comment, please address in a separate commit since they are not necessarily related to this PR but to a uncrustify version update.

**Comment by fjmolinas** [comment]
> @kaspar030 @leandrolanzieri @maribu : it went green tada
> 
> The restructuring implements following logic of timer choice:
> 
>     * if module `ztimer_usec` is used use the (xtimer) `ztimer_periph_timer` for it.
> 
>     * if module `ztimer_msec` is used and `ztrimer_periph_rtt` use `ztimer_periph_rtt` for it else use the (xtimer) `ztimer_periph_timer` for it.
> 
>     * if module `ztimer_sec` is used and `ztimer_periph_rtc` is used use `ztimer_periph_rtc` for it else if `ztrimer_periph_rtt` used, use `ztimer_periph_rtt` for it else use the (xtimer) `ztimer_periph_timer` for it.
> 
>     * Non of the `ztimer_*sec` modules depend on each other they only depend on `ztimer_periph*`, assuming `ztimer_periph_timer` is always available they can always be implemented (the: `No suitable ZTIMER_MSEC config. Maybe add USEMODULE += ztimer_usec?` error message is gone the replacement message should never be triggert since there is always a `ztimer_periph_timer` atm).

Exactly what would be nice to document other than in the ifdef mess.

File: sys/ztimer/auto_init.c (hunk 9 above)
Comment (kfessel) [reply sys/ztimer/auto_init.c:93]: The rtt-frequency is set by the rtt initialization which is mcu dependent, if it can be set, a number of boards default to 1Hz (see # #16116 i think this should be changend but this is not part of this PR) therefore the Test for the rtt-frequency. @maribu recently made the sugestion in #16116 that periph timer might be useable for rtt-/lp-timers which would make the configuration availible to ztimer.

Comment (fjmolinas) [reply sys/ztimer/auto_init.c:93]: Now that I think of it, shifting only works for power of two frequencies., so this is fine.

**Comment by kfessel** [comment]
>     * it would be nice if the choice of initialization and backend was documented in another way as only with ifdefs, since they are really hard to read.

K, i will add some inline comments  that will describe what is done in the ifdef, I think the new ifdef-hell is a large improvement over the one that is there without the PR

>     * can we add distinctive tests for the different timers Since the application is the same mostly, I think an approach as the one used for the network driver compile test can be used

I will have a look at this maybe I need some help by @maribu

>     * the history needs fixing :)

You realy dont like my take on history :) -- i think there will be a rebase at the end
 
>     * there are some uncrustify comment, please address in a separate commit since they are not necessarily related to this PR but to a uncrustify version update.

uncrustify also does not like me  :) -- i will address that

**Comment by fjmolinas** [comment]
> You realy dont like my take on history :) -- i think there will be a rebase at the end

No, sorry if it came across that way, its just that since you don't use the `--fixup` workflow its hard to follow what commits are being patched, and therefore when I look at the individual commits I see some weird things. But this, as you say, will probably be fixed with the rebase, it justs helps when commits follow the contributing guidelines [fixups](https://github.com/RIOT-OS/RIOT/blob/master/CONTRIBUTING.md#add-fixup-commits-during-review) and [commits](https://github.com/RIOT-OS/RIOT/blob/master/CONTRIBUTING.md#add-fixup-commits-during-review)

**Comment by kfessel** [comment]
the fixup also adds the comments
@fjmolinas: you are right the "rename reorder uncrustify" is a fixup but i also think its title is better matching (and much easyser to see in qgit fixup can be anything

**Comment by kfessel** [comment]
uncrustfy is just plain wrong wrong about sys/include/ztimer.h#L297 and sys/include/ztimer.h#L309  its against the riot uncrustfy rules

and no one in their right mind would consider to do that
```
--- a/sys/include/ztimer/config.h
+++ b/sys/include/ztimer/config.h
@@ -109,10 +109,10 @@ extern "C" {
 #ifndef CONFIG_ZTIMER_TIMER_REQUIRED_PM_MODE
 #  ifdef CONFIG_ZTIMER_USEC_REQUIRED_PM_MODE
 #    define CONFIG_ZTIMER_TIMER_REQUIRED_PM_MODE \
-            CONFIG_ZTIMER_USEC_REQUIRED_PM_MODE
+    CONFIG_ZTIMER_USEC_REQUIRED_PM_MODE
 #  else
 #    define CONFIG_ZTIMER_TIMER_REQUIRED_PM_MODE \
-            ZTIMER_CLOCK_NO_REQUIRED_PM_MODE
+    ZTIMER_CLOCK_NO_REQUIRED_PM_MODE
 #  endif
 #endif
 ```
it kills the last bit of readablity that was left (after a line breake that it sugested there for 2-5 chars

**Comment by kaspar030** [comment]
> uncrustfy is just plain wrong wrong about sys/include/ztimer.h#L297 and sys/include/ztimer.h#L309 its against the riot uncrustfy rules

uncrustify is currently only warning. just use up-to-date uncrustify and ignore the warnings.

> 

> it kills the last bit of readablity that was left (after a line breake that it sugested there for 2-5 chars

main reason to use uncrustify is to not deal with formatting anymore but still have consistent looks. right now, that might be at the cost of some readability sometimes.

**Comment by maribu** [comment]
Maybe we could configure `uncrustify` to never ever suggest a line break. We have other tools already complaining about long lines and I agree that `uncrustify` really does a poor job at breaking lines.

**Comment by kaspar030** [comment]
> Maybe we could configure `uncrustify` to never ever suggest a line break. We have other tools already complaining about long lines and I agree that `uncrustify` really does a poor job at breaking lines.

I do like being able to do "strg-f" (my shortcut for uncrustify) and be done with it, and enjoy not going through a CI static-test fail/look at/manually fix cycle...

We should definitely update the uncrustify on CI. if it takes too long to go to 2020.04 LTS, maybe self-compile the latest version.

**Comment by kfessel** [comment]
but even the version that comes with 2020.10 wants to make that define less readable

ATM i think changing CONFIG_ZTIMER_TIMER_REQUIRED_PM_MODE to CONFIG_ZTIMER_TIM_REQ_PM_MODE  to cicumvent that uncrustify problem

**Comment by maribu** [comment]
> I do like being able to do "strg-f" (my shortcut for uncrustify) and be done with it, and enjoy not going through a CI static-test fail/look at/manually fix cycle...

Add `set cc=101` to your `.vimrc` / `init.vim` and you'll never accidentally surpass the limit.

Also: `uncrustify` still is at 80 chars (not 100). Fixing this would solve the issue at hand.

But generally: The coding convention only recommends to keep lines shorter than up to 100 chars - as otherwise this sets and insensitive to use short but unreadable names.

**Comment by kfessel** [comment]
> > I do like being able to do "strg-f" (my shortcut for uncrustify) and be done with it, and enjoy not going through a CI static-test fail/look at/manually fix cycle...
> 
> Add `set cc=101` to your `.vimrc` / `init.vim` and you'll never accidentally surpass the limit.
> 
> Also: `uncrustify` still is at 80 chars (not 100). Fixing this would solve the issue at hand.
> 
> But generally: The coding convention only recommends to keep lines shorter than up to 100 chars - as otherwise this sets and insensitive to use short but unreadable names.

so i remove the linebreakes and ignore the uncrustify?

**Comment by kaspar030** [comment]
> ATM i think changing CONFIG_ZTIMER_TIMER_REQUIRED_PM_MODE to CONFIG_ZTIMER_TIM_REQ_PM_MODE to cicumvent that uncrustify problem

you can also use code-style-ignore comments.

**Comment by maribu** [comment]
> so i remove the linebreakes and ignore the uncrustify?

Let's wait a bit. I'll PR a change to 100 chars. That should go in without discussion, as our coding convention was updated to allow this and we just forgot to also touch uncrustify. Maybe the issue is gone with that already.

**Comment by kfessel** [comment]
#16149 will update riot-uncrustify to 100 chars (riot-coding standard)

**Comment by kfessel** [comment]
just checked the complains by uncrustify:
-it is wrong about wrong about sys/include/ztimer.h#L297 and sys/include/ztimer.h#L309 , changing to suggestion breaks the rule written in config (there is no multipication in that line)
-sys/include/ztimer/config.h#L108 #L119 #L130 are complient with riot coding convention of upto 100 chars in a line
breaking them hinders readability (especially if the uncrustify suggestion is used)

**Comment by kfessel** [comment]
rebase for uncrustify config update

**Comment by maribu** [comment]
At this point I'm still formally author of this PR, but haven't really authored much of the changes. IMO it would make sense if I close the PR and @kfessel opens a new one with the current state (which references this one). This would IMO better document the development history and authorship.

**Comment by kfessel** [comment]
Please continue the review in #16172
